util/audio.py

```diff
@@ -133,3 +210,115 @@ def vad_split(audio_frames,
         yield b''.join(voiced_frames), \
               frame_duration_ms * (frame_index - len(voiced_frames)), \
               frame_duration_ms * (frame_index + 1)
+
+
+def pack_number(n, num_bytes):
+    return n.to_bytes(num_bytes, 'big', signed=False)
+
+
+def unpack_number(data):
+    return int.from_bytes(data, 'big', signed=False)
+
+
+def get_opus_frame_size(rate):
+    return 60 * rate // 1000
+
+
+def write_opus(opus_file, audio_format, audio_data):
```

Comment:
The `opuslib` is just a codec (it requires a container format) - this is what this code does. I kinda made up this simplistic format that will never be used outside SDBs on single-file basis.

Reply:
RIght, but shouldn't it be easier to maintain by just re-using some opus library to do the job ?

Reply:
There is no single library for that. Just tools that put Opus into a multi-purpose container (e.g. Ogg) and are bigger and more complicated dependencies. I 100% understand your doubt about this, but doing this would definitely rise complexity without any other win than "being compatible" **within** a format that is anyhow completely in our ownership (SDB).

And it's not really complicated - the "format" is just:
- a **header** that consists of the PCM length followed by the audio-format (rate, channels, width) and
- **frames** that consist of a chunk size indicator followed by the chunk bytes for the codec.

That's it.

Reply:
> There is no single library for that. Just tools that put Opus into a multi-purpose container (e.g. Ogg) and are bigger and more complicated dependencies. I 100% understand your doubt about this, but doing this would definitely rise complexity without any other win than "being compatible" **within** a format that is anyhow completely in our ownership (SDB).
> 
> And it's not really complicated - the "format" is just:
> 
>     * a **header** that consists of the PCM length followed by the audio-format (rate, channels, width) and
> 
>     * **frames** that consist of a chunk size indicator followed by the chunk bytes for the codec.
> 
> 
> That's it.

The complicated part for me was the re-invention of a container. But given those constraints, I agree it's okay.

Reply:
It's too late now, but for future documentation and better understanding of the requirements, did you consider using tar? Do we need extra stuff from the tar archive format doesn't offer?

Reply:
I am glad you asked this, as this will make a great rationale-section in the format description:
Actually I already added TAR export to the DSAlign project and this was the biggest motivation to go SDB:
- TAR is not offering any random access - consequences:
  - no playing / looking-up at a given index
  - no bucketing for better/performant external sorting
- TAR entries would have to be in an own format or container (to distinguish audio and transcript)
- The file-structure (naming) plus content types (Opus/UTF-8) would also build some kind of standard where the TAR format is just file-name provider and separator

Reply:
It is surprisingly hard to get encoding plus container-packaging as a simple Python solution. Most solutions are either wrappers of `ffmpeg` (no viable option, as starting another process per sample is too slow) or bindings to its `libavformat`:

> This power does come with some responsibility as working with media is horrendously complicated and PyAV can't abstract it away or make all the best decisions for you. If the ffmpeg command does the job without you bending over backwards, PyAV is likely going to be more of a hindrance than a help.

Both approaches also add additional dependencies on top of the anyhow required `libopus`.
Another (cleaner) option is implementing the `Ogg` container format directly in Python. `Ogg` is not super complex, but it would add some code to maintain.

Reply:
Just one missing aspect for consideration: The `Ogg` header comes with some space-wise and computation-wise (checksum) overhead.

Reply:
@tilmankamp Space wise what is this overhead? 50%? 1%? 0.01?

Reply:
We all synced on this and decided to keep the custom container approach (with a custom mime-type).